### ISSUE TITLE
[Test container]When running tests getting private service fails

### ISSUE THREAD

**Issue by Mpreyzner** [symfony/symfony#28528]
**Symfony version(s) affected**: 4.1.4, 4.1.3

**Description**  
When running tests getting private service fails
According to this: https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing it should be now possible.

Running bin/phpunit results in:
> Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The "App\Service\Service" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

After changing service visibility it works as intended.

**How to reproduce**  
Run bin/phpunit in repository with BugReproducer - https://github.com/Mpreyzner/symfony_visibility_bug

**Possible Solution**  

**Additional context**

**Comment by nicolas-grekas** [comment]
Only non-removed and non-inlined services can be fetched. If you can't get it in 4.1, it means the service has been removed because it's not used anywhere. So why test it?

**Comment by lukaszwit** [comment]
I've got similar issue. And my service is definitely used. According to info from compiler pass my service (the on I would like to fetch from container in test) is not private nor public. Is that ok?

**Comment by kejwmen** [comment]
> what do mean by inline service?

Service can be inlined when some [conditions are met](https://github.com/symfony/dependency-injection/blob/master/Compiler/InlineServiceDefinitionsPass.php#L158). 
If I understood it correctly, usual scenario is: when service is not public and is injected into one service, then it can be inlined. As you can see in related compiler pass this is more complicated.

Could you share result of `bin/console debug:container RI --env=test` with us?

By the way: I know it's probably not related, but when you test specific implementation (`R`) fetch it explicitly from container instead of fetching `RInterface` - the result will be the same, but you fetch exactly what you want to test.

**Comment by DonCallisto** [comment]
I don't agree with this: if I'm developing app via TDD (as correctly stated above) this should be retrievable without any alias.

Moreover there always be an "outer" service that you'll use in controllers (for instance) so, pratically, it would be used in my application.
Injecting it in controller, btw, does not seems to work as service will be gone anyway.

So my question is: what *private* service can be fetched that isn't an unused service? Can anyone bring a valid example? I'm thinking about this scenario without finding a proper answer.

**Comment by tecnocat** [comment]
Same here, we have a very simply test like this:

```
    /**
     * @return void
     */
    public function testFindById(): void
    {
        self::bootKernel();
        $dummyRepository = self::$container->get(DummyRepositoryInterface::class);
        $dummy           = $dummyRepository->findById(1);
        $this->assertInstanceOf(Dummy::class, $dummy);
        $this->assertNull($dummy->getDeletedAt());
    }

```

The test was working well until the last composer update in the last week, something was changed in DI Symfony component...

Using the implementation (DummyRepository) instead of an interface (DummyRepositoryInterface) for the test to working again broken the SOLID principle.

**Comment by stof** [comment]
@DonCallisto the only way to achieve that would be to disable all the optimization of the container inlining some services. But that would mean that the tests are not running against the same container than the actual logic (and in case of circular object graphs, this can make a huge difference, as some circular references are allowed as long as the cycle does not reach a public service in the meantime (which would prevent reorganizing the instantiation of the whole cyclic graph) for instance.

**Comment by DonCallisto** [comment]
@stof I know what you mean, but at this point, what are the advantages of using this kind of test container? Only to retrieve directly from container those services? Just asking, your answer is very clear and I strongly agree with you.

**Comment by iamhappycoder** [comment]
used anywhere = DI tree starts at a controller

**Comment by mnapoli** [comment]
I just stumbled across this problem myself, trying to test a service that is not used for now in the rest of the codebase (refactoring ongoing, the new service is not used for now but I still want to test it).

**Comment by andrecadete** [comment]
Kinda sounds a bit anti-TDDish approach to "not allow" testing Objects simply because they are not in use yet. Any TDD purist will be upset / concerned with this.

**Comment by thomaskonrad** [comment]
Same here. I'm in a situation where I offer a service in a bundle, but the service is not used in the bundle itself. It might or might not be used somewhere else in the project. This means that I cannot test the service in a `KernelTestCase` independently, because the service will get removed from the container.

Why not make the test container also contain unused services?

**Comment by xabbuh** [comment]
IMO your test case then should reflect the use case you try to test. Thus, register a service in your test environment that uses the service your bundle provides.
